### PR TITLE
Split Dataset into train/test/val

### PR DESCRIPTION
**Pull request by dreadatour**
See related issue, docstrings and tests.

##### Train-test split

```python
from datachain import DataChain
from datachain.toolkit import train_test_split

# Load a DataChain from a storage source (e.g., S3 bucket)
dc = DataChain.from_storage("s3://bucket/dir/")

# Perform a 70/30 train-test split
train, test = train_test_split(dc, [0.7, 0.3])

# Save the resulting splits
train.save("dataset_train")
test.save("dataset_test")
```

##### Train-test-validation split

```python
train, test, val = train_test_split(dc, [0.7, 0.2, 0.1])
train.save("dataset_train")
test.save("dataset_test")
val.save("dataset_val")
```

### TODO

- `seed`
- `shuffle`

### CODE REVIEW

File: src/datachain/toolkit/split.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        - The weights do not need to sum to 1; they will be normalized internally.
+          For example:
+            - `[0.7, 0.3]` corresponds to a 70/30 split;
+            - `[2, 1, 1]` corresponds to a 50/25/25 split;
+    """
+    if len(weights) < 2:
+        raise ValueError("Weights should have at least two elements")
+
+    weights_normalized = [round(weight * 1000 / sum(weights)) for weight in weights]
+    print(weights_normalized)
```
Comment (0x2b3bfa0) [inline src/datachain/toolkit/split.py:57]: Is this `print` intentional?

Comment (dreadatour) [reply src/datachain/toolkit/split.py]: 🤫

File: src/datachain/toolkit/split.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        - The splits are random but deterministic, based on Dataset `sys__rand` field.
+        - The weights do not need to sum to 1; they will be normalized internally.
+          For example:
+            - `[0.7, 0.3]` corresponds to a 70/30 split;
+            - `[2, 1, 1]` corresponds to a 50/25/25 split;
+    """
+    if len(weights) < 2:
+        raise ValueError("Weights should have at least two elements")
+
+    weights_normalized = [round(weight * 1000 / sum(weights)) for weight in weights]
```
Comment (0x2b3bfa0) [inline src/datachain/toolkit/split.py:56]: Minor nitpick: consider making these values truly "normalized" in range 0-1, and moving the `* 1000` below, closer to the `%` operation.

**Review by 0x2b3bfa0 — APPROVED** [review]

File: src/datachain/toolkit/split.py (hunk 2 above)
Comment (0x2b3bfa0) [reply src/datachain/toolkit/split.py:56]: Also, I have been playing a bit with this normalization approach and sometimes I have found unexpected results:

```python
>>> sum(normalize([5, 5]))
1000
>>> sum(normalize([3, 3, 1]))
1001  # What!?
```

Not sure how important is that for this use case (maybe it's not), but looks like a potential subtle bug to me.

Comment (dreadatour) [reply src/datachain/toolkit/split.py]: I guess this is because of `round`. Have it removed.

Comment (dreadatour) [reply src/datachain/toolkit/split.py]:
```
In [1]: weights = [3, 3, 1]

In [2]: [weight / sum(weights) for weight in weights]
Out[2]: [0.42857142857142855, 0.42857142857142855, 0.14285714285714285]

In [3]: sum([weight / sum(weights) for weight in weights])
Out[3]: 1.0
```

**Review by shcheklein — APPROVED** [review]
Amazing, thanks @dreadatour

File: src/datachain/toolkit/split.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        raise ValueError("Weights should have at least two elements")
+    if any(weight < 0 for weight in weights):
+        raise ValueError("Weights should be non-negative")
+
+    weights_normalized = [weight / sum(weights) for weight in weights]
+
+    return [
+        dc.filter(
+            C("sys__rand") % 1000 >= round(sum(weights_normalized[:index]) * 1000),
+            C("sys__rand") % 1000 < round(sum(weights_normalized[: index + 1]) * 1000),
```
Comment (dmpetrov) [inline src/datachain/toolkit/split.py:64]: Please do not use double underscore. It's technical detail, not user facing API.

File: src/datachain/toolkit/split.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    if len(weights) < 2:
+        raise ValueError("Weights should have at least two elements")
+    if any(weight < 0 for weight in weights):
+        raise ValueError("Weights should be non-negative")
+
+    weights_normalized = [weight / sum(weights) for weight in weights]
+
+    return [
+        dc.filter(
+            C("sys__rand") % 1000 >= round(sum(weights_normalized[:index]) * 1000),
```
Comment (dmpetrov) [inline src/datachain/toolkit/split.py:63]: should it be a bigger number like 100K to cover more precise ratios like [99.007 0.001, 0.002]?

**Review by dmpetrov — COMMENTED** [review]
Amazing!!! Just a couple of comments are inline.

File: src/datachain/toolkit/split.py (hunk 4 above)
Comment (0x2b3bfa0) [reply src/datachain/toolkit/split.py:63]: Yes! 💯 If we know the exact bounds of `sys__rand` (do we?) we should use them to provide as much precision as possible.

Comment (dmpetrov) [reply src/datachain/toolkit/split.py:63]: it's a random Int, so IntMax ~2B. I'd not be too crazy about this and make it like 100K or 1M.

Comment (0x2b3bfa0) [reply src/datachain/toolkit/split.py:63]: If it's a 32-bit ([un]signed?) integer, we can perfectly cover the entire range. 🤷‍♂️

Comment (dmpetrov) [reply src/datachain/toolkit/split.py:63]: yeah, bit it might depend on architecture/DB that can potentially complicate the logic. If not - sure.

Comment (0x2b3bfa0) [reply src/datachain/toolkit/split.py:63]: Moved here: https://github.com/iterative/datachain/pull/611
